src/routes/tools/degree-days/_data.js

```diff
@@ -0,0 +1,239 @@
+ * raster series and add extra props. These props describe how the data
+ * should be fetched from the API or how the data is shown in the charts.
+ * List of extra props:
+ * slugs - list of raster series names in API
+ * mark - describes how the timeseries will be displayed in the chart (line/area)
+ * visible - controls whether line/area in chart is hidden/shown
+ *      when user clicks on corresponding legend key. Note: The stats component
+ *      also uses this prop to include/remove series from summary calculations
+ * @param {object} config
+ * @return {array}
+ */
+
+// The observed data is usually a single raster series, so only 1 slug
+const getObservedSeries = ({ climvarId }) => {
+  return OBSERVED.map((d) => {
+    const slugs = [`${climvarId}_day_${d.id}`];
+    return { ...d, slugs, mark: "line", visible: true };
+  });
+};
+
+// For each model, there are usually 2 raster series in the API,
+// the modeled historical (1950-2005) and modeled projections (2006-2099/2021)
+const getModelSeries = ({ climvarId, scenarioId, modelIds, indicatorId }) => {
+  return PRIORITY_10_MODELS.filter((d) => modelIds.includes(d.id)).map((d) => {
+    const slugs = [
+      // `${climvarId}_day_${d.id}_historical`,
+      `${climvarId}_day_${d.id}_${scenarioId}`,
+    ];
+    return { ...d, slugs, mark: "line", visible: true, indicatorId };
+  });
+};
+
+// The ensemble has to be assembled from the max and min of 10/all models
+const getEnsembleSeries = ({ scenarioId, indicatorId }) => {
+  return ENSEMBLES.filter((d) => d.id === `${scenarioId}_range`).map((d) => {
+    const slugs = [
+      `${indicatorId}_year_ens32avg_historical`,
+      `${indicatorId}_year_ens32avg_${scenarioId}`,
+    ];
+    return { ...d, slugs, mark: "area", visible: true };
+  });
+};
+
+// determines next URL path segment following the slug
+const urlPathAfterSlug = (indicatorId, isEnsemble) => {
+  if (["cdd", "hdd"].includes(indicatorId) && !isEnsemble) return indicatorId;
+  return "events";
+};
+
+const filterReponseByMonths = (response, monthIds) =>
+  response.filter(({ date }) => monthIds.has(date.getMonth()));
+
+const aggregateMonthlyData = (response) =>
+  rollups(
+    response,
+    (arr) => +format(".0f")(mean(arr, (d) => d.value)),
+    (dd) => dd.date.getFullYear()
+  ).map(([year, value]) => ({ date: new Date(Date.UTC(year)), value }));
+
+/**
+ * Fetches data from the events endpoint in Cal-Adapt API
+ * Input parameters:
+ * slug - name of raster series in API
+ * params - object with props for geometry, stat, units, etc.
+ * method - default is GET, POST used for user uploaded boundaries
+ * @param {object}
+ * @return {array}
+ */
+const fetchEvents = async ({
+  slug,
+  params,
+  method = "GET",
+  indicatorId,
+  monthIds,
+  isEnsemble,
+}) => {
+  const url = `${apiEndpoint}/series/${slug}/${urlPathAfterSlug(
+    indicatorId,
+    isEnsemble
+  )}/`;
+  const [response, error] = await handleXHR(fetchData(url, params, method));
```

Comment:
On thinking some more, we probably don't need to use `handleXHR` here. I wrote that function to track errors when users upload their own boundary for aggregation (because there are a whole bunch of different error messages returned by the API). But now all that checking first happens in the `tools/partials/ChangeLocation`, so by the time the geometry reaches `fetchEvents` it's geometry that's good to go. Instead we can just check for 404 errors (incorrect slugs) or empty arrays. If the array is empty we can show the No Data state.

Reply:
So swap it and use `window.fetch`? 

For the empty data state we will need to rethink the logic in ExploreData. Currently it will show the loading spinner if `$dataStore` is an empty array. We might just want to hide the loading after `$dataStore` changes to no longer be `null`. Or better yet, have a separate `isLoadingData` Svelte Store that is a boolean which is toggled to `true` before a request is made and then toggled to `false` after a request has finished or errored. That way we can show a loading spinner whenever data is fetched, but that might not be necessary if data has already been cached by the browser.

Reply:
Adding a separate isLoadingData store is a good idea.

Reply:
Cool. The one thing to be cautious of from a UX perspective is to not flash a loading spinner very quickly after the settings change. It can be triggering and also make the data update seem like it's taking longer than it is.